File: processing/syntactic/expressions/atomic/constant_expr.py
from __future__ import annotations

from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.token import Token
from processing.location import span_marker
from processing.syntactic.expressions.expr import Expr
from processing.syntactic.expressions.groups import AtomExpression


class ConstantExpr(AtomExpression):
    """ const_expr:
        CONST_TOKEN;
    """

    def __init__(
            self,
            parent: Expr = None,
            literal: Token = None
    ):
        super().__init__(parent)
        self.literal = literal

    @span_marker
    def parse(self) -> ConstantExpr:
        self.literal = self.stream.eat_any()
        return self

    def to_axion(self, c: CodeBuilder):
        c += self.literal

    def to_csharp(self, c: CodeBuilder):
        c += self.literal

    def to_python(self, c: CodeBuilder):
        c += self.literal

File: Axion.Core/Processing/Syntactic/Expressions/Atomic/ConstantExpr.cs
using Axion.Core.Processing.CodeGen;
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Common;
using Axion.Core.Processing.Syntactic.Expressions.TypeNames;
using Axion.Core.Processing.Traversal;

namespace Axion.Core.Processing.Syntactic.Expressions.Atomic {
    /// <summary>
    ///     <c>
    ///         const-expr:
    ///             CONST-TOKEN | STRING+;
    ///     </c>
    /// </summary>
    public class ConstantExpr : AtomExpr {
        public Token Literal { get; set; }

        [NoTraversePath]
        public override TypeName ValueType => Literal.ValueType;

        public static ConstantExpr ParseNew(Expr parent) {
            return new ConstantExpr(parent).Parse();
        }

        public ConstantExpr(
            Expr   parent,
            Token? literal = null
        ) : base(parent) {
            Literal = literal;
        }

        public ConstantExpr(
            Expr   parent,
            string literal
        ) : base(parent) {
            Literal = new Token(Source, value: literal);
        }

        public ConstantExpr Parse() {
            SetSpan(() => { Literal = Stream.EatAny(); });
            return this;
        }

        public override void ToDefault(CodeWriter c) {
            c.Write(Literal);
        }
    }
}